### PR TITLE
GEODE-9484: Improve sending message to multy destinations

### PR DESCRIPTION
**Pull request by mivanac**
Problem:
When replicating to other servers, current logic would first create connections toward all servers, and then replicate data over those connections. Creation of connections will last, until all connections are created, or destination server/s is/are declared dead.

Solution:
New proposal is, when replicating data, first try to create connections to all destinations in one attempt. For all created connections we will replicate data. After this step is completed, we will try to create remaining connections to all unreachable destinations, until connections are established, or destination is declared dead. If connections are established we will replicate data.

For more info see https://cwiki.apache.org/confluence/display/GEODE/Improve+data+inconsistency+in+replicated+regions

### For all changes:
- [*] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [*] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [*] Is your initial contribution a single, squashed commit?

- [*] Does `gradlew build` run cleanly?

- [x] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### CODE REVIEW

**Review by onichols-pivotal — COMMENTED** [review]
_comment deleted_

**Review by onichols-pivotal — COMMENTED** [review]
_comment deleted_

**Comment by mivanac** [comment]
Hi @Bill , @agingade , @echobravopapa , @kirklund and @kamilla1201.
Can you review this PR.

Thanks

**Comment by mivanac** [comment]
Hi @Bill , @agingade , @echobravopapa , @kirklund and @kamilla1201.
Can you review this PR.

Thanks

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/UpdatePropagationDUnitTest.java [hunk 1]
```diff
@@ -91,30 +99,63 @@ public final void postSetUp() throws Exception {
...
+    hostnameServer1 = NetworkUtils.getServerHostName(server1.getHost());
+    hostnameServer3 = NetworkUtils.getServerHostName(server3.getHost());
 
     IgnoredException.addIgnoredException("java.net.SocketException");
     IgnoredException.addIgnoredException("Unexpected IOException");
   }
 
+
+
+  @Test
```
Comment (pivotal-jbarrett) [inline geode-core/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/UpdatePropagationDUnitTest.java:124]: Please rename the test to the current naming convention, `UpdatePropogationDistributedTest`.

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/UpdatePropagationDUnitTest.java [hunk 2]
```diff
@@ -91,30 +99,63 @@ public final void postSetUp() throws Exception {
     // Server2 VM
     server2 = host.getVM(1);
 
+    // Server3 VM
```
Comment (pivotal-jbarrett) [inline geode-core/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/UpdatePropagationDUnitTest.java:102]: These comments don't provide any more detail than the variable name, so they can just go away.

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/UpdatePropagationDUnitTest.java [hunk 3]
```diff
@@ -91,30 +99,63 @@ public final void postSetUp() throws Exception {
...
+  public void updatesArePropagatedToAllMembersWhenOneKilled() throws Exception {
+    client1.invoke(
+        () -> createClientCache(hostnameServer1, PORT1));
+    client2.invoke(
+        () -> createClientCache(hostnameServer3, PORT3));
+    int entries = 20;
+    AsyncInvocation invocation = client1.invokeAsync(() -> doPuts(entries));
+
+    // Wait for some entries to be put
+    Thread.sleep(5000);
```
Comment (pivotal-jbarrett) [inline geode-core/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/UpdatePropagationDUnitTest.java:134]: This should be an awaited assertion rather than arbitrary time. Sleeps like this eventually result in flaky behavior under different system loads.

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/UpdatePropagationDUnitTest.java [hunk 4]
```diff
@@ -305,6 +358,32 @@ private void verifyUpdates() {
...
+      } catch (Exception e) {
+        // ignore
+      }
+      Thread.sleep(1000);
+    }
+  }
+
+  private int getNotNullEntriesNumber(int entries) {
+    int notNullEntries = 0;
+    Region r1 = getCache().getRegion(SEPARATOR + REGION_NAME);
```
Comment (pivotal-jbarrett) [inline geode-core/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/UpdatePropagationDUnitTest.java:376]: Don't use raw types. Please add type params.

File: geode-core/src/main/java/org/apache/geode/distributed/internal/direct/DirectChannel.java [hunk 5]
```diff
@@ -458,8 +472,14 @@ private ConnectExceptions getConnections(Membership mgr, DistributionMessage msg
           if (ackTimeout > 0) {
             startTime = System.currentTimeMillis();
           }
-          Connection con = conduit.getConnection(destination, preserveOrder, retry, startTime,
-              ackTimeout, ackSDTimeout);
+          Connection con;
```
Comment (pivotal-jbarrett) [inline geode-core/src/main/java/org/apache/geode/distributed/internal/direct/DirectChannel.java:475]: Can this be `final`?
Perhaps this if block should be extracted into its own method for clarity.
I wouldn't mind a rename of `con` to `connection` while we are here.

File: geode-core/src/main/java/org/apache/geode/internal/tcp/Connection.java [hunk 6]
```diff
@@ -1021,7 +1021,9 @@ static Connection createSender(final Membership<InternalDistributedMember> mgr,
             // do not change the text of this exception - it is looked for in exception handlers
             throw new IOException("Cannot form connection to alert listener " + remoteAddr);
           }
-
+          if (onlyOneTry) {
```
Comment (pivotal-jbarrett) [inline geode-core/src/main/java/org/apache/geode/internal/tcp/Connection.java:1024]: is `onlyOneTry` really "fail on first try" or "do not retry"?

File: geode-core/src/main/java/org/apache/geode/internal/tcp/ConnectionTable.java [hunk 7]
```diff
@@ -400,6 +401,10 @@ private Connection getSharedConnection(InternalDistributedMember id, boolean sch
           throw new IOException("Cannot form connection to alert listener " + id);
         }
 
+        if (onlyOneTry) {
```
Comment (pivotal-jbarrett) [inline geode-core/src/main/java/org/apache/geode/internal/tcp/ConnectionTable.java:404]: This doesn't feel like an appropriately named variable here. It feels more like a "do not wait for connection".

File: geode-core/src/main/java/org/apache/geode/internal/tcp/TCPConduit.java [hunk 8]
```diff
@@ -909,6 +906,104 @@ public Connection getConnection(InternalDistributedMember memberAddress,
...
+   */
+  public Connection getFirstScanForConnection(InternalDistributedMember memberAddress,
+      final boolean preserveOrder, long startTime, long ackTimeout,
+      long ackSATimeout) throws IOException, DistributedSystemDisconnectedException {
+    if (stopped) {
+      throw new DistributedSystemDisconnectedException("The conduit is stopped");
+    }
+
+    InternalDistributedMember memberInTrouble = null;
+    Connection conn = null;
```
Comment (pivotal-jbarrett) [inline geode-core/src/main/java/org/apache/geode/internal/tcp/TCPConduit.java:931]: Please don't use abbreviated variable names.

**Review by pivotal-jbarrett — CHANGES_REQUESTED** [review]

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/UpdatePropagationDUnitTest.java (hunk 1 above)
Comment (mivanac) [reply geode-core/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/UpdatePropagationDUnitTest.java:122]: Thanks for comments. Updated

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/UpdatePropagationDUnitTest.java (hunk 2 above)
Comment (mivanac) [reply geode-core/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/UpdatePropagationDUnitTest.java]: Updated

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/UpdatePropagationDUnitTest.java (hunk 3 above)
Comment (mivanac) [reply geode-core/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/UpdatePropagationDUnitTest.java]: Updated

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/UpdatePropagationDUnitTest.java (hunk 4 above)
Comment (mivanac) [reply geode-core/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/UpdatePropagationDUnitTest.java]: Updated

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/UpdatePropagationDistributedTest.java [hunk 9]
```diff
@@ -358,22 +356,31 @@ private void verifyUpdates() {
...
       } catch (Exception e) {
-        // ignore
       }
       Thread.sleep(1000);
     }
   }
 
+  private boolean getRegionSizeGreaterThen(int numEntries) {
+    Region<String, String> r1 = getCache().getRegion(SEPARATOR + REGION_NAME);
+    assertNotNull(r1);
```
Comment (pivotal-jbarrett) [inline geode-core/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/UpdatePropagationDistributedTest.java:377]: Use AssertJ here.

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/UpdatePropagationDistributedTest.java [hunk 10]
```diff
@@ -358,22 +356,31 @@ private void verifyUpdates() {
     });
   }
 
+  private void verifyMinEntriesInserted() {
+    await().until(() -> getRegionSizeGreaterThen(minNumEntries));
```
Comment (pivotal-jbarrett) [inline geode-core/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/UpdatePropagationDistributedTest.java:360]: Try:
```java
await().untilAsserted(() ->
  assertThat(getCache().getRegion(SEPARATOR + REGION_NAME)).hasSizeGreaterThan(minNumEntries));
```

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/UpdatePropagationDistributedTest.java [hunk 11]
```diff
@@ -358,22 +356,31 @@ private void verifyUpdates() {
...
+    Region<String, String> r1 = getCache().getRegion(SEPARATOR + REGION_NAME);
+    assertNotNull(r1);
+    return (r1.size() > numEntries);
+  }
+
   private int getNotNullEntriesNumber(int entries) {
     int notNullEntries = 0;
-    Region r1 = getCache().getRegion(SEPARATOR + REGION_NAME);
+    Region<String, String> r1 = getCache().getRegion(SEPARATOR + REGION_NAME);
     assertNotNull(r1);
```
Comment (pivotal-jbarrett) [inline geode-core/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/UpdatePropagationDistributedTest.java:384]: Please replace all with AssertJ.

**Review by pivotal-jbarrett — CHANGES_REQUESTED** [review]

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/UpdatePropagationDistributedTest.java (hunk 9 above)
Comment (mivanac) [reply geode-core/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/UpdatePropagationDistributedTest.java]: Updated

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/UpdatePropagationDistributedTest.java (hunk 10 above)
Comment (mivanac) [reply geode-core/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/UpdatePropagationDistributedTest.java]: Thanks for comments. Updated.

File: geode-core/src/main/java/org/apache/geode/internal/tcp/TCPConduit.java [hunk 12]
```diff
@@ -777,18 +776,16 @@ public Connection getConnection(InternalDistributedMember memberAddress,
...
-              // ignored
+          try {
+            if (logger.isDebugEnabled()) {
+              logger.debug("Closing old connection.  conn={} before retrying. memberInTrouble={}",
+                  conn, memberInTrouble);
             }
+            conn.closeForReconnect("closing before retrying");
+          } catch (CancelException ex) {
+            throw ex;
+          } catch (Exception ex) {
```
Comment (pivotal-jbarrett) [inline geode-core/src/main/java/org/apache/geode/internal/tcp/TCPConduit.java:787]: We can eliminate warnings here by use the `ignored` variable name rather than comment.
```java
} catch (Exception ignored) {
}
```

File: geode-core/src/main/java/org/apache/geode/internal/tcp/TCPConduit.java [hunk 13]
```diff
@@ -909,6 +906,104 @@ public Connection getConnection(InternalDistributedMember memberAddress,
...
+      long ackSATimeout) throws IOException, DistributedSystemDisconnectedException {
+    if (stopped) {
+      throw new DistributedSystemDisconnectedException("The conduit is stopped");
+    }
+
+    Connection connection = null;
+    stopper.checkCancelInProgress(null);
+    boolean interrupted = Thread.interrupted();
+    try {
+      // If this is the second time through this loop, we had problems.
```
Comment (pivotal-jbarrett) [inline geode-core/src/main/java/org/apache/geode/internal/tcp/TCPConduit.java:934]: What loop?

File: geode-core/src/main/java/org/apache/geode/internal/tcp/TCPConduit.java [hunk 14]
```diff
@@ -909,6 +906,104 @@ public Connection getConnection(InternalDistributedMember memberAddress,
...
+                    memberAddress));
+          } else if (connection.isClosing()
+              || !connection.getRemoteAddress().equals(memberAddress)) {
+            if (logger.isDebugEnabled()) {
+              logger.debug("Got an old connection for {}: {}@{}", memberAddress, connection,
+                  connection.hashCode());
+            }
+            connection.closeOldConnection("closing old connection");
+            connection = null;
+            retryForOldConnection = true;
```
Comment (pivotal-jbarrett) [inline geode-core/src/main/java/org/apache/geode/internal/tcp/TCPConduit.java:960]: I am rarely a fan of do/while loops and this one tracking this exit value was initially hard to follow. I think a infinite loop with throw exceptions and early breaks shortens it and clears up the purpose.
```java
final debugEnabled = logger.isDebugEnabled();
while(true) {
  connection = getConTable()...;
  if (connection == null) {
    throw new IOException(...);
  }
  if (!(connection.isClosing() || ..)) {
    break;
  }
  if (debugEnabled) {
    logger.debug(...);
  }
  connection.closeOldConnection(...);
}
```
Throwing the exception gets handled below the same as having set the `problem` variable.
You don't need to reset the `connection` to `null` after it is closed since the next statement to execute sets it anyway.

I would even go one step further and move this whole loop logic into a method.
```java
final Connection = getConnectionThatIsNotClosed(...);
```
```java
@NotNull Connection getConnectionThatIsNotClosed(...) {
  final debugEnabled = logger.isDebugEnabled();
  while(true) {
    final Connection connection = getConTable()...;
    if (connection == null) {
      throw new IOException(...);
    }
    if (!(connection.isClosing() || ..)) {
      return connection;
    }
    if (debugEnabled) {
      logger.debug(...);
    }
    connection.closeOldConnection(...);
  }
}
```

File: geode-core/src/main/java/org/apache/geode/internal/tcp/TCPConduit.java [hunk 15]
```diff
@@ -909,6 +906,104 @@ public Connection getConnection(InternalDistributedMember memberAddress,
...
+            debugRetry = true;
+          }
+        } while (retryForOldConnection);
+        if (debugRetry && logger.isDebugEnabled()) {
```
Comment (pivotal-jbarrett) [inline geode-core/src/main/java/org/apache/geode/internal/tcp/TCPConduit.java:964]: Does this log message do anything for us?

**Review by pivotal-jbarrett — CHANGES_REQUESTED** [review]

File: geode-core/src/main/java/org/apache/geode/internal/tcp/TCPConduit.java (hunk 14 above)
Comment (pivotal-jbarrett) [reply geode-core/src/main/java/org/apache/geode/internal/tcp/TCPConduit.java]: Ah, that makes sense.

**Review by pivotal-jbarrett — CHANGES_REQUESTED** [review]

**Review by pivotal-jbarrett — APPROVED** [review]
